File: LogFilter.cs
// LogFilter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using System.Threading;

class Config
{
    [JsonPropertyName("filters")] public List<string> Filters { get; set; } = new List<string>();
    [JsonPropertyName("levels")] public List<string> Levels { get; set; } = new List<string>();
    [JsonPropertyName("since")] public string Since { get; set; } = "1h";
}

class LogEntry
{
    public DateTime? Timestamp { get; set; }
    public string Level { get; set; }
    public string Message { get; set; }
    public string Raw { get; set; }

    public LogEntry(string raw)
    {
        Raw = raw;
        Parse(raw);
    }

    private void Parse(string line)
    {
        string[] patterns = {
            @"\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\]\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)",
            @"(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)",
            @"(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)",
            @"^(ERROR|WARN|INFO|DEBUG)\s+(.*)"
        };
        foreach (var pattern in patterns)
        {
            var match = Regex.Match(line, pattern);
            if (match.Success)
            {
                if (match.Groups.Count == 4)
                {
                    if (DateTime.TryParse(match.Groups[1].Value, out DateTime ts))
                        Timestamp = ts;
                    Level = match.Groups[2].Value;
                    Message = match.Groups[3].Value;
                }
                else if (match.Groups.Count == 3)
                {
                    Level = match.Groups[1].Value;
                    Message = match.Groups[2].Value;
                }
                return;
            }
        }
        Message = line;
    }
}

class LogFilter
{
    private static readonly string ConfigFile = "log_filter_config.json";
    private static readonly Dictionary<string, string> Colors = new Dictionary<string, string>
    {
        {"red", "\x1b[91m"}, {"green", "\x1b[92m"}, {"yellow", "\x1b[93m"},
        {"blue", "\x1b[94m"}, {"cyan", "\x1b[96m"}, {"reset", "\x1b[0m"}
    };
    private static readonly Dictionary<string, string> LevelColors = new Dictionary<string, string>
    {
        {"ERROR", "red"}, {"WARN", "yellow"}, {"INFO", "green"}, {"DEBUG", "cyan"}
    };

    private Config config;

    public LogFilter() { config = LoadConfig(); }

    private Config LoadConfig()
    {
        if (!File.Exists(ConfigFile)) return new Config();
        string json = File.ReadAllText(ConfigFile);
        return JsonSerializer.Deserialize<Config>(json) ?? new Config();
    }

    private void SaveConfig()
    {
        string json = JsonSerializer.Serialize(config, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(ConfigFile, json);
    }

    private long ParseSince(string sinceStr)
    {
        if (string.IsNullOrEmpty(sinceStr)) return 3600;
        int num = int.Parse(sinceStr.Substring(0, sinceStr.Length - 1));
        char unit = sinceStr[^1];
        switch (unit)
        {
            case 'm': return num * 60;
            case 'h': return num * 3600;
            case 'd': return num * 86400;
            default: return 3600;
        }
    }

    private string Highlight(string text, string filter, string regexPattern)
    {
        if (!string.IsNullOrEmpty(filter))
            return text.Replace(filter, Colors["yellow"] + filter + Colors["reset"]);
        if (!string.IsNullOrEmpty(regexPattern))
            return Regex.Replace(text, regexPattern, Colors["yellow"] + "$0" + Colors["reset"]);
        foreach (var kv in LevelColors)
        {
            if (text.Contains(kv.Key))
                return text.Replace(kv.Key, Colors[kv.Value] + kv.Key + Colors["reset"]);
        }
        return text;
    }

    public void ProcessFile(string filepath, string filter, string regexPattern, List<string> levels,
                            string sinceStr, int context, bool watch, bool stats, bool noColor)
    {
        if (!File.Exists(filepath)) { Console.WriteLine($"File '{filepath}' not found."); return; }

        if (stats) { ShowStats(filepath); return; }

        long sinceSec = ParseSince(string.IsNullOrEmpty(sinceStr) ? config.Since : sinceStr);
        DateTime cutoff = DateTime.Now.AddSeconds(-sinceSec);

        var lines = File.ReadAllLines(filepath);
        var entries = lines.Select(l => new LogEntry(l)).ToList();

        var filtered = new List<(int idx, string line, LogEntry entry)>();
        for (int i = 0; i < entries.Count; i++)
        {
            var entry = entries[i];
            var line = lines[i];
            if (entry.Timestamp.HasValue && entry.Timestamp.Value < cutoff) continue;
            if (levels.Any() && !levels.Contains(entry.Level)) continue;
            if (!string.IsNullOrEmpty(filter) && !line.Contains(filter, StringComparison.OrdinalIgnoreCase)) continue;
            if (!string.IsNullOrEmpty(regexPattern) && !Regex.IsMatch(line, regexPattern)) continue;
            filtered.Add((i, line, entry));
        }

        Console.WriteLine($"\n📜 Log Viewer – {filepath}");
        if (!string.IsNullOrEmpty(filter) || !string.IsNullOrEmpty(regexPattern))
            Console.WriteLine($"Filter: {filter ?? regexPattern}");
        if (levels.Any()) Console.WriteLine($"Levels: {string.Join(", ", levels)}");
        Console.WriteLine($"Watch mode: {watch}\n");

        if (context > 0)
        {
            var shown = new HashSet<int>();
            foreach (var f in filtered)
                for (int offset = -context; offset <= context; offset++)
                {
                    int idx = f.idx + offset;
                    if (idx >= 0 && idx < lines.Length) shown.Add(idx);
                }
            foreach (int idx in shown.OrderBy(i => i))
            {
                string prefix = filtered.Any(f => f.idx == idx) ? "> " : "  ";
                string line = lines[idx];
                if (!noColor) line = Highlight(line, filter, regexPattern);
                Console.WriteLine($"{prefix}{idx+1,4}: {line}");
            }
        }
        else
        {
            foreach (var f in filtered)
            {
                string line = f.line;
                if (!noColor) line = Highlight(line, filter, regexPattern);
                Console.WriteLine($"{f.idx+1,4}: {line}");
            }
        }
        Console.WriteLine($"\n{filtered.Count} lines matched.");

        if (watch) WatchMode(filepath, filter, regexPattern, levels, sinceSec, context, noColor);
    }

    private void WatchMode(string filepath, string filter, string regexPattern, List<string> levels,
                           long sinceSec, int context, bool noColor)
    {
        Console.WriteLine("\n👁️  Watching for new lines (press Ctrl+C to stop)...");
        long lastPos = new FileInfo(filepath).Length;
        while (true)
        {
            Thread.Sleep(1000);
            var info = new FileInfo(filepath);
            if (info.Length > lastPos)
            {
                using var fs = new FileStream(filepath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
                fs.Seek(lastPos, SeekOrigin.Begin);
                using var reader = new StreamReader(fs);
                string line;
                while ((line = reader.ReadLine()) != null)
                {
                    var entry = new LogEntry(line);
                    if (entry.Timestamp.HasValue && entry.Timestamp.Value < DateTime.Now.AddSeconds(-sinceSec)) continue;
                    if (levels.Any() && !levels.Contains(entry.Level)) continue;
                    if (!string.IsNullOrEmpty(filter) && !line.Contains(filter, StringComparison.OrdinalIgnoreCase)) continue;
                    if (!string.IsNullOrEmpty(regexPattern) && !Regex.IsMatch(line, regexPattern)) continue;
                    if (!noColor) line = Highlight(line, filter, regexPattern);
                    Console.WriteLine($"  {line}");
                }
                lastPos = fs.Position;
            }
        }
    }

    private void ShowStats(string filepath)
    {
        var counts = new Dictionary<string, int> { {"ERROR", 0}, {"WARN", 0}, {"INFO", 0}, {"DEBUG", 0}, {"OTHER", 0} };
        foreach (var line in File.ReadLines(filepath))
        {
            var entry = new LogEntry(line);
            if (entry.Level != null && counts.ContainsKey(entry.Level))
                counts[entry.Level]++;
            else if (!string.IsNullOrWhiteSpace(line))
                counts["OTHER"]++;
        }
        Console.WriteLine("\n📊 Log Statistics:");
        foreach (var level in new[] { "ERROR", "WARN", "INFO", "DEBUG", "OTHER" })
        {
            string color = LevelColors.GetValueOrDefault(level, "reset");
            Console.WriteLine($"  {Colors[color]}{level,5}{Colors["reset"]}: {counts[level]}");
        }
    }

    static void Main(string[] args)
    {
        if (args.Length < 1) { Console.WriteLine("Usage: LogFilter <file> [options]"); return; }
        var app = new LogFilter();
        var parsed = ParseArgs(args);
        string filepath = args[0];

        if (parsed.ContainsKey("save"))
        {
            var cfg = app.config;
            if (parsed.ContainsKey("filter")) cfg.Filters = new List<string> { parsed["filter"] };
            if (parsed.ContainsKey("level")) cfg.Levels = parsed["level"].Split(',').ToList();
            if (parsed.ContainsKey("since")) cfg.Since = parsed["since"];
            app.SaveConfig();
            Console.WriteLine("✅ Filter settings saved.");
            return;
        }

        string filter = parsed.GetValueOrDefault("filter");
        string regex = parsed.GetValueOrDefault("regex");
        List<string> levels = parsed.ContainsKey("level") ?
            parsed["level"].Split(',').ToList() : new List<string>();
        string since = parsed.GetValueOrDefault("since");
        int context = parsed.ContainsKey("context") ? int.Parse(parsed["context"]) : 0;
        bool watch = parsed.ContainsKey("watch");
        bool stats = parsed.ContainsKey("stats");
        bool noColor = parsed.ContainsKey("no-color");

        // Use config defaults
        if (string.IsNullOrEmpty(filter) && string.IsNullOrEmpty(regex) && !levels.Any() && string.IsNullOrEmpty(since))
        {
            if (app.config.Filters.Any()) filter = app.config.Filters[0];
            levels = app.config.Levels;
            since = app.config.Since;
        }

        app.ProcessFile(filepath, filter, regex, levels, since, context, watch, stats, noColor);
    }

    static Dictionary<string, string> ParseArgs(string[] args)
    {
        var dict = new Dictionary<string, string>();
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i].StartsWith("--"))
            {
                string key = args[i].Substring(2);
                if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                    dict[key] = args[++i];
                else
                    dict[key] = "";
            }
        }
        return dict;
    }
}

File: LogFilter.java
// LogFilter.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import java.util.regex.*;
import java.util.stream.*;
import com.google.gson.*;

class Config {
    List<String> filters = new ArrayList<>();
    List<String> levels = new ArrayList<>();
    String since = "1h";
}

class LogEntry {
    LocalDateTime timestamp;
    String level;
    String message;
    String raw;

    LogEntry(String raw) {
        this.raw = raw;
        parse(raw);
    }

    private void parse(String line) {
        String[] patterns = {
            "\\[(\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2})\\]\\s+(ERROR|WARN|INFO|DEBUG)\\s+(.*)",
            "(\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2})\\s+(ERROR|WARN|INFO|DEBUG)\\s+(.*)",
            "(\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2})\\s+(ERROR|WARN|INFO|DEBUG)\\s+(.*)",
            "^(ERROR|WARN|INFO|DEBUG)\\s+(.*)"
        };
        for (String pattern : patterns) {
            Pattern p = Pattern.compile(pattern);
            Matcher m = p.matcher(line);
            if (m.find()) {
                if (m.groupCount() == 3) {
                    try {
                        this.timestamp = LocalDateTime.parse(m.group(1), DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
                    } catch (DateTimeParseException e) {
                        try {
                            this.timestamp = LocalDateTime.parse(m.group(1), DateTimeFormatter.ISO_DATE_TIME);
                        } catch (DateTimeParseException e2) {}
                    }
                    this.level = m.group(2);
                    this.message = m.group(3);
                } else if (m.groupCount() == 2) {
                    this.level = m.group(1);
                    this.message = m.group(2);
                }
                return;
            }
        }
        this.message = line;
    }
}

public class LogFilter {
    private static final String CONFIG_FILE = "log_filter_config.json";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private static final Map<String, String> COLORS = new HashMap<>();
    private static final Map<String, String> LEVEL_COLORS = new HashMap<>();
    static {
        COLORS.put("red", "\033[91m");
        COLORS.put("green", "\033[92m");
        COLORS.put("yellow", "\033[93m");
        COLORS.put("blue", "\033[94m");
        COLORS.put("cyan", "\033[96m");
        COLORS.put("reset", "\033[0m");
        LEVEL_COLORS.put("ERROR", "red");
        LEVEL_COLORS.put("WARN", "yellow");
        LEVEL_COLORS.put("INFO", "green");
        LEVEL_COLORS.put("DEBUG", "cyan");
    }

    private Config config;

    public LogFilter() { this.config = loadConfig(); }

    private Config loadConfig() {
        try {
            Path path = Paths.get(CONFIG_FILE);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                return gson.fromJson(json, Config.class);
            }
        } catch (Exception e) {}
        return new Config();
    }

    private void saveConfig() {
        try {
            Files.write(Paths.get(CONFIG_FILE), gson.toJson(config).getBytes());
        } catch (Exception e) {}
    }

    private long parseSince(String sinceStr) {
        if (sinceStr == null || sinceStr.isEmpty()) return 3600;
        int num = Integer.parseInt(sinceStr.substring(0, sinceStr.length()-1));
        char unit = sinceStr.charAt(sinceStr.length()-1);
        switch (unit) {
            case 'm': return num * 60;
            case 'h': return num * 3600;
            case 'd': return num * 86400;
            default: return 3600;
        }
    }

    private String highlight(String text, String filter, String regexPattern) {
        if (filter != null && !filter.isEmpty()) {
            return text.replace(filter, COLORS.get("yellow") + filter + COLORS.get("reset"));
        }
        if (regexPattern != null && !regexPattern.isEmpty()) {
            return text.replaceAll(regexPattern, COLORS.get("yellow") + "$0" + COLORS.get("reset"));
        }
        for (Map.Entry<String, String> entry : LEVEL_COLORS.entrySet()) {
            if (text.contains(entry.getKey())) {
                return text.replace(entry.getKey(), COLORS.get(entry.getValue()) + entry.getKey() + COLORS.get("reset"));
            }
        }
        return text;
    }

    private void showStats(String filepath) throws IOException {
        Map<String, Integer> counts = new HashMap<>();
        counts.put("ERROR", 0); counts.put("WARN", 0);
        counts.put("INFO", 0); counts.put("DEBUG", 0); counts.put("OTHER", 0);
        List<String> lines = Files.readAllLines(Paths.get(filepath));
        for (String line : lines) {
            LogEntry entry = new LogEntry(line);
            if (entry.level != null && counts.containsKey(entry.level)) {
                counts.put(entry.level, counts.get(entry.level) + 1);
            } else if (!line.trim().isEmpty()) {
                counts.put("OTHER", counts.get("OTHER") + 1);
            }
        }
        System.out.println("\n📊 Log Statistics:");
        for (String level : Arrays.asList("ERROR", "WARN", "INFO", "DEBUG", "OTHER")) {
            String color = LEVEL_COLORS.getOrDefault(level, "reset");
            System.out.printf("  %s%5s%s: %d%n", COLORS.get(color), level, COLORS.get("reset"), counts.get(level));
        }
    }

    public void processFile(String filepath, String filter, String regexPattern,
                            List<String> levels, String sinceStr, int context,
                            boolean watch, boolean stats, boolean noColor) throws IOException, InterruptedException {
        Path path = Paths.get(filepath);
        if (!Files.exists(path)) {
            System.err.println("File '" + filepath + "' not found.");
            return;
        }

        if (stats) {
            showStats(filepath);
            return;
        }

        long sinceSec = parseSince(sinceStr != null ? sinceStr : config.since);
        LocalDateTime cutoff = LocalDateTime.now().minusSeconds(sinceSec);

        List<String> lines = Files.readAllLines(path);
        List<LogEntry> entries = lines.stream().map(LogEntry::new).collect(Collectors.toList());

        List<Map<String, Object>> filtered = new ArrayList<>();
        for (int i = 0; i < entries.size(); i++) {
            LogEntry entry = entries.get(i);
            String line = lines.get(i);
            // Time filter
            if (entry.timestamp != null && entry.timestamp.isBefore(cutoff)) continue;
            // Level filter
            if (!levels.isEmpty() && !levels.contains(entry.level)) continue;
            // Pattern filter
            if (filter != null && !line.toLowerCase().contains(filter.toLowerCase())) continue;
            if (regexPattern != null && !line.matches(".*" + regexPattern + ".*")) continue;
            Map<String, Object> map = new HashMap<>();
            map.put("idx", i);
            map.put("line", line);
            map.put("entry", entry);
            filtered.add(map);
        }

        System.out.printf("\n📜 Log Viewer – %s%n", filepath);
        if (filter != null || regexPattern != null) System.out.printf("Filter: %s%n", filter != null ? filter : regexPattern);
        if (!levels.isEmpty()) System.out.printf("Levels: %s%n", String.join(", ", levels));
        System.out.printf("Watch mode: %s%n%n", watch ? "on" : "off");

        if (context > 0) {
            Set<Integer> shown = new HashSet<>();
            for (Map<String, Object> f : filtered) {
                int idx = (int)f.get("idx");
                for (int offset = -context; offset <= context; offset++) {
                    int n = idx + offset;
                    if (n >= 0 && n < lines.size()) shown.add(n);
                }
            }
            List<Integer> sorted = new ArrayList<>(shown);
            Collections.sort(sorted);
            for (int idx : sorted) {
                String prefix = filtered.stream().anyMatch(f -> (int)f.get("idx") == idx) ? "> " : "  ";
                String line = lines.get(idx);
                if (!noColor) line = highlight(line, filter, regexPattern);
                System.out.printf("%s%4d: %s%n", prefix, idx+1, line);
            }
        } else {
            for (Map<String, Object> f : filtered) {
                int idx = (int)f.get("idx");
                String line = (String)f.get("line");
                if (!noColor) line = highlight(line, filter, regexPattern);
                System.out.printf("%4d: %s%n", idx+1, line);
            }
        }
        System.out.printf("%n%d lines matched.%n", filtered.size());

        if (watch) {
            watchMode(filepath, filter, regexPattern, levels, sinceSec, context, noColor);
        }
    }

    private void watchMode(String filepath, String filter, String regexPattern,
                           List<String> levels, long sinceSec, int context, boolean noColor) throws IOException, InterruptedException {
        System.out.println("\n👁️  Watching for new lines (press Ctrl+C to stop)...");
        long lastPos = Files.size(Paths.get(filepath));
        while (true) {
            Thread.sleep(1000);
            long newSize = Files.size(Paths.get(filepath));
            if (newSize > lastPos) {
                try (RandomAccessFile raf = new RandomAccessFile(filepath, "r")) {
                    raf.seek(lastPos);
                    String line;
                    while ((line = raf.readLine()) != null) {
                        LogEntry entry = new LogEntry(line);
                        if (entry.timestamp != null && entry.timestamp.isBefore(LocalDateTime.now().minusSeconds(sinceSec))) continue;
                        if (!levels.isEmpty() && !levels.contains(entry.level)) continue;
                        if (filter != null && !line.toLowerCase().contains(filter.toLowerCase())) continue;
                        if (regexPattern != null && !line.matches(".*" + regexPattern + ".*")) continue;
                        if (!noColor) line = highlight(line, filter, regexPattern);
                        System.out.printf("  %s%n", line);
                    }
                    lastPos = raf.getFilePointer();
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: LogFilter <file> [options]");
            return;
        }
        LogFilter app = new LogFilter();
        String filepath = args[0];
        Map<String, String> params = new HashMap<>();
        for (int i = 1; i < args.length; i++) {
            if (args[i].startsWith("--") && i+1 < args.length) {
                params.put(args[i].substring(2), args[++i]);
            } else if (args[i].startsWith("--")) {
                params.put(args[i].substring(2), "");
            }
        }

        if (params.containsKey("save")) {
            Config cfg = app.config;
            if (params.containsKey("filter")) cfg.filters = Arrays.asList(params.get("filter"));
            if (params.containsKey("level")) cfg.levels = Arrays.asList(params.get("level").split(","));
            if (params.containsKey("since")) cfg.since = params.get("since");
            app.saveConfig();
            System.out.println("✅ Filter settings saved.");
            return;
        }

        String filter = params.get("filter");
        String regex = params.get("regex");
        List<String> levels = params.containsKey("level") ?
            Arrays.asList(params.get("level").split(",")) : new ArrayList<>();
        String since = params.get("since");
        int context = params.containsKey("context") ? Integer.parseInt(params.get("context")) : 0;
        boolean watch = params.containsKey("watch");
        boolean stats = params.containsKey("stats");
        boolean noColor = params.containsKey("no-color");

        // Use config defaults
        if (filter == null && regex == null && levels.isEmpty() && since == null) {
            if (!app.config.filters.isEmpty()) filter = app.config.filters.get(0);
            levels = app.config.levels;
            since = app.config.since;
        }

        app.processFile(filepath, filter, regex, levels, since, context, watch, stats, noColor);
    }
}
